File: bus/bus.py
import time
import asyncio
from typing import Optional

from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from httpx import AsyncClient
from pydantic import BaseModel

app = FastAPI()
app.add_middleware(
    CORSMiddleware,
    allow_origins=[
        "https://peter.demin.dev",
        "http://localhost",
        "http://localhost:8080",
    ],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
SCHEDULE_URL = "https://s3.amazonaws.com/kcm-alerts-realtime-prod/tripupdates_pb.json"


class Schedule:
    def __init__(self, cache_ttl: int = 60) -> None:
        self._updated_at = 0
        self._data = None
        self._ok = False
        self._cache_ttl = cache_ttl

    async def get(self) -> tuple[bool, dict]:
        if not self._data or self._updated_at < time.time() - self._cache_ttl:
            self._ok, self._data = await self._fetch()
            self._updated_at = time.time()
        return self._ok, self._data

    @staticmethod
    async def _fetch() -> tuple[bool, dict]:
        async with AsyncClient() as ac:
            response = await ac.get(SCHEDULE_URL)
        if response.status_code != 200:
            return False, {
                "error": "Tripupdates API returned non-200",
                "details": response.content,
            }
        return True, response.json()


class ScheduleAPI:
    def __init__(self, schedule: Schedule) -> None:
        self._schedule = schedule

    async def stop_times(self, stop_ids: str):
        ok, data = await self._schedule.get()
        if not ok:
            return {
                "error": "Tripupdates API returned non-200",
                "details": data,
            }
        return {
            "arrivals": {
                stop_id: self._extact_arrival_times(data, stop_id)
                for stop_id in stop_ids.split(",")
            },
            "routes": {
                stop_id: self._extact_routes(data, stop_id)
                for stop_id in stop_ids.split(",")
            },
        }

    def _extact_arrival_times(self, data: dict, stop_id: str) -> list[int]:
        arrivals = []
        for entity in data["entity"]:
            for stop_time_update in entity["trip_update"]["stop_time_update"]:
                if stop_time_update["stop_id"] == stop_id:
                    arrivals.append(stop_time_update["arrival"]["time"])
        return arrivals

    def _extact_routes(self, data: dict, stop_id: str) -> dict[int, list[str]]:
        routes = {}
        for entity in data["entity"]:
            arrivals = []
            for stop_time_update in entity["trip_update"]["stop_time_update"]:
                if stop_time_update["stop_id"] == stop_id:
                    arrivals.append(stop_time_update["arrival"]["time"])
            if arrivals:
                route_id = entity["trip_update"]["trip"]["route_id"]
                routes.setdefault(route_id, []).extend(arrivals)
        return {route_id: sorted(times) for route_id, times in routes.items()}

    def register(self, app: FastAPI, prefix: str = "") -> None:
        app.add_api_route(prefix + "/{stop_ids}", self.stop_times, methods=["GET"])


class Metadata(BaseModel):
    name: str
    location: str
    start_ts: float
    finish_ts: float
    context: str = ""
    input_hint: str = ""
    output_hint: str = ""


class Checkpoint(BaseModel):
    input: str
    output: str
    metadata: Metadata


class CheckpointRepository:
    def add(self, tenant: str, checkpoint: Checkpoint) -> None:
        raise NotImplementedError()

    def fetch_all(self, tenant: str) -> list[Checkpoint]:
        raise NotImplementedError()


class InMemoryCheckpointRepository(CheckpointRepository):
    DEFAULT_CHECKPOINT_LIMIT = 100
    DEFAULT_TTL = 60.0  # seconds

    def __init__(
        self, checkpoint_limit: int = DEFAULT_CHECKPOINT_LIMIT, ttl: float = DEFAULT_TTL
    ) -> None:
        self._storage: dict[str, list[tuple[float, Checkpoint]]] = {}
        self._checkpoint_limit = checkpoint_limit
        self._ttl = ttl

    def add(self, tenant: str, checkpoint: Checkpoint) -> None:
        self._trim_old(tenant)
        if len(self._storage[tenant]) > self._checkpoint_limit:
            self._storage[tenant] = self._storage[tenant][-self._checkpoint_limit:]
        self._storage[tenant].append((time.time(), checkpoint))

    def fetch_all(self, tenant: str) -> list[Checkpoint]:
        self._trim_old(tenant)
        return [v for _, v in self._storage[tenant]]

    def _trim_old(self, tenant: str) -> None:
        now = time.time()
        self._storage[tenant] = [
            (t, v) for t, v in (self._storage.get(tenant) or []) if now - t < self._ttl
        ]


class CheckpointAPI:
    def __init__(self, checkpoint_repository: CheckpointRepository) -> None:
        self._checkpoint_repository = checkpoint_repository

    async def ingest_checkpoint(self, tenant: str, checkpoint: Checkpoint) -> dict:
        self._checkpoint_repository.add(tenant, checkpoint)
        asyncio.sleep(0.1)
        return {"status": "ok"}

    async def get_checkpoints(self, tenant: str) -> list[dict]:
        return [
            checkpoint.dict()
            for checkpoint in self._checkpoint_repository.fetch_all(tenant)
        ]

    def register(self, app: FastAPI, prefix: str = "") -> None:
        app.add_api_route(
            prefix + "/{tenant}", self.ingest_checkpoint, methods=["POST"]
        )
        app.add_api_route(prefix + "/{tenant}", self.get_checkpoints, methods=["GET"])


ScheduleAPI(Schedule()).register(app, "/bus")
checkpoint_api = CheckpointAPI(InMemoryCheckpointRepository())
checkpoint_api.register(app, "/i8t")

File: bus/bus.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"os"
	"os/signal"
	"sort"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"
)

const scheduleURL = "https://s3.amazonaws.com/kcm-alerts-realtime-prod/tripupdates_pb.json"

// ---- Minimal structs matching only the fields we use ----
type TripUpdates struct {
	Entity []struct {
		TripUpdate struct {
			Trip struct {
				RouteID string `json:"route_id"`
			} `json:"trip"`
			StopTimeUpdate []struct {
				StopID  string `json:"stop_id"`
				Arrival struct {
					Time int64 `json:"time"`
				} `json:"arrival"`
			} `json:"stop_time_update"`
		} `json:"trip_update"`
	} `json:"entity"`
}

// ---- Cache ----
type Cache struct {
	mu        sync.RWMutex
	expiresAt time.Time
	ttl       time.Duration
	ok        bool
	data      *TripUpdates
	details   []byte // raw response body in case of error
}

func NewCache(ttl time.Duration) *Cache { return &Cache{ttl: ttl} }

func (c *Cache) Get(client *http.Client) (bool, *TripUpdates, []byte) {
	now := time.Now()
	c.mu.Lock()
	defer c.mu.Unlock()
	fresh := c.ok == true && c.data != nil && now.Before(c.expiresAt)
	if !fresh {
		c.ok, c.data, c.details = fetchTripUpdates(client)
		c.expiresAt = time.Now().Add(c.ttl)
	}
	return c.ok, c.data, c.details
}

func fetchTripUpdates(client *http.Client) (bool, *TripUpdates, []byte) {
	req, _ := http.NewRequest("GET", scheduleURL, nil)
	req.Header.Set("User-Agent", "https://peter.demin.dev/12_articles/57-bus-time.html")
	resp, err := client.Do(req)
	if err != nil {
		return false, nil, []byte(err.Error())
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != http.StatusOK {
		return false, nil, body
	}
	var tu TripUpdates
	if err := json.Unmarshal(body, &tu); err != nil {
		// If JSON is bad, surface the raw body
		return false, nil, body
	}
	return true, &tu, nil
}

// ---- API layer ----
type ScheduleAPI struct {
	cache  *Cache
	client *http.Client
}

func NewScheduleAPI(ttl time.Duration) *ScheduleAPI {
	return &ScheduleAPI{
		cache:  NewCache(ttl),
		client: &http.Client{Timeout: 10 * time.Second},
	}
}

type responsePayload struct {
	Arrivals map[string][]int64            `json:"arrivals"`
	Routes   map[string]map[string][]int64 `json:"routes"`
}

func (s *ScheduleAPI) stopTimesHandler(w http.ResponseWriter, r *http.Request) {
	// CORS preflight handled globally; this path handles GET.
	stopIDsParam := strings.TrimPrefix(r.URL.Path, "/bus/")
	stopIDs := splitCSV(stopIDsParam)

	ok, data, details := s.cache.Get(s.client)
	w.Header().Set("Content-Type", "application/json")

	if !ok || data == nil {
		_ = json.NewEncoder(w).Encode(map[string]any{
			"error":   "Tripupdates API returned non-200",
			"details": string(details),
		})
		return
	}

	result := responsePayload{
		Arrivals: make(map[string][]int64, len(stopIDs)),
		Routes:   make(map[string]map[string][]int64, len(stopIDs)),
	}

	for _, stopID := range stopIDs {
		result.Arrivals[stopID] = extractArrivalTimes(data, stopID)

		routes := extractRoutes(data, stopID)
		// sort each route’s times
		for _, times := range routes {
			sort.Slice(times, func(i, j int) bool { return times[i] < times[j] })
		}
		result.Routes[stopID] = routes
	}

	_ = json.NewEncoder(w).Encode(result)
}

func extractArrivalTimes(data *TripUpdates, stopID string) []int64 {
	var arrivals []int64
	for _, e := range data.Entity {
		for _, stu := range e.TripUpdate.StopTimeUpdate {
			if stu.StopID == stopID {
				arrivals = append(arrivals, stu.Arrival.Time)
			}
		}
	}
	return arrivals
}

func extractRoutes(data *TripUpdates, stopID string) map[string][]int64 {
	routes := make(map[string][]int64)
	for _, e := range data.Entity {
		var times []int64
		for _, stu := range e.TripUpdate.StopTimeUpdate {
			if stu.StopID == stopID {
				times = append(times, stu.Arrival.Time)
			}
		}
		if len(times) > 0 {
			routes[e.TripUpdate.Trip.RouteID] = append(routes[e.TripUpdate.Trip.RouteID], times...)
		}
	}
	return routes
}

func splitCSV(s string) []string {
	parts := strings.Split(s, ",")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}

// ---- CORS middleware (origins & credentials like your FastAPI setup) ----
var allowedOrigins = map[string]struct{}{
	"https://peter.demin.dev": {},
	"http://localhost":        {},
	"http://localhost:8080":   {},
}

func withCORS(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		origin := r.Header.Get("Origin")
		if _, ok := allowedOrigins[origin]; ok {
			w.Header().Set("Access-Control-Allow-Origin", origin)
			w.Header().Set("Vary", "Origin")
			w.Header().Set("Access-Control-Allow-Credentials", "true")
			w.Header().Set("Access-Control-Allow-Methods", "GET, OPTIONS")
			w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")
		}
		if r.Method == http.MethodOptions {
			w.WriteHeader(http.StatusNoContent)
			return
		}
		next.ServeHTTP(w, r)
	})
}

func systemdListener() (net.Listener, error) {
	// Verify SystemD socket activation works:
	lp := os.Getenv("LISTEN_PID")
	lf := os.Getenv("LISTEN_FDS")
	if lp == "" || lf == "" {
		return nil, fmt.Errorf("not socket-activated: LISTEN_PID/LISTEN_FDS missing")
	}
	pid, err := strconv.Atoi(lp)
	if err != nil || pid != os.Getpid() {
		return nil, fmt.Errorf("LISTEN_PID=%q does not match pid=%d", lp, os.Getpid())
	}
	n, err := strconv.Atoi(lf)
	if err != nil || n < 1 {
		return nil, fmt.Errorf("LISTEN_FDS=%q invalid", lf)
	}
	if n != 1 {
		return nil, fmt.Errorf("expected exactly 1 listening fd, got %d", n)
	}
	_ = os.Unsetenv("LISTEN_PID")
	_ = os.Unsetenv("LISTEN_FDS")
	_ = os.Unsetenv("LISTEN_FDNAMES")

	// systemd hands us the first FD at 3
	f := os.NewFile(uintptr(3), "systemd-listener")
	if f == nil {
		return nil, fmt.Errorf("failed to open fd 3")
	}
	// net.FileListener dup's the FD internally; close our *os.File handle afterwards.
	ln, err := net.FileListener(f)
	_ = f.Close()
	if err != nil {
		return nil, fmt.Errorf("net.FileListener(fd=3): %w", err)
	}
	return ln, nil
}

func main() {
	var (
		addr = flag.String("addr", "", "TCP listen address, instead of SystemD socket activation")
	)
	flag.Parse()

	api := NewScheduleAPI(60 * time.Second)

	mux := http.NewServeMux()
	mux.HandleFunc("/bus/", api.stopTimesHandler)
	mux.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte("ok"))
	})

	srv := &http.Server{
		Handler:           withCORS(mux),
		ReadHeaderTimeout: 5 * time.Second,
	}

	// Graceful shutdown context
	ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
	defer stop()

	// Serve on TCP or UNIX socket
	if *addr != "" {
		srv.Addr = *addr
		log.Printf("listening on %s (tcp)", *addr)
		go func() {
			if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
				log.Fatalf("ListenAndServe: %v", err)
			}
		}()
	} else {
		// SystemD activation socket
		ln, err := systemdListener()
		if err != nil {
			log.Fatalf("listen unix: %v", err)
		}
		log.Printf("listening on SystemD activation socket")
		go func() {
			if err := srv.Serve(ln); err != nil && err != http.ErrServerClosed {
				log.Fatalf("Serve(socket): %v", err)
			}
		}()
	}

	// Wait for signal, then graceful shutdown
	<-ctx.Done()
	stop()
	shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if err := srv.Shutdown(shutdownCtx); err != nil {
		log.Printf("graceful shutdown error: %v", err)
	}
	log.Println("server stopped")
}
